QgsGeometry: add support for area3D

ptitjano opened the pull request:
## Description

`area3D` is the 3-dimensional surface area of the geometry.

For a Point or a LineString, it is always equal to 0.
For a Polygon, it is equal to the 3D area of the exterior ring minus
the 3d area of the interior rings.
For a Polyhedral and a TIN, it is equal to the sum of the 3d areas of
all its patches.
For a GeometryCollection, it is equal to the sum of the 3d areas of all
its geometries.

One simply need to define the computation for a Polygon and then, all
the other geometries can be deduced.

To compute the 3D area of a Polygon, a local planar coordinate system
is built from three points lying on the polygon’s plane, and all the
rings are projected onto this plane. Then, the signed area of each
projected ring is computed. The area is computed by by adding the
exterior ring and subtracting the interior rings.

This needs #64206 and #64201 to merged first and thus it includes those PRs.

## Update 19 December 

The area3D computation is now based on 3d shoelace formula for QgsLineString as suggested by @rouault 
The QgsCircularString::sumUpArea3D still uses a 2D projection.

ptitjano committed:
qgsabstractgeometry: Add support for area3D

`area3D` is the 3-dimensional surface area of the geometry.

For a Point or a LineString, it is always equal to 0.
For a Polygon, it is equal to the 3D area of the exterior ring minus
the 3d area of the interior rings.
For a Polyhedral and a TIN, it is equal to the sum of the 3d areas of
all its patches.
For a GeometryCollection, it is equal to the sum of the 3d areas of all
its geometries.

One simply need to define the computation for a Polygon and then, all
the other geometries can be deduced.

To compute the 3D area of a Polygon, `QgsCurve::sumUpArea3D()` is
used. For a closed ring, its absolute value is equal to the 3d area of
the ring.

ptitjano committed:
qgsgeometry: Add support for area3D

This simply calls area3D on the underlying `QgsAbstractGeometry`.

nyalldawson commented on src/core/geometry/qgscurvepolygon.h:
```diff
@@ -470,6 +472,13 @@ class CORE_EXPORT QgsCurvePolygon: public QgsSurface
+     *
+     * \since QGIS 4.0
+     */
+    bool hasPlane3D( QgsPoint &ptA, QgsPoint &ptB, QgsPoint &ptC, double epsilon = std::numeric_limits<double>::epsilon() ) const;
```
```suggestion
    bool has3DPlane( QgsPoint &ptA SIP_OUT, QgsPoint &ptB SIP_OUT, QgsPoint &ptC SIP_OUT, double epsilon = std::numeric_limits<double>::epsilon() ) const;
```

nyalldawson commented on src/core/geometry/qgscurvepolygon.cpp:
```diff
@@ -492,6 +492,70 @@ double QgsCurvePolygon::area() const
+
+  /*
+   * compute the area for each ring in the local basis
+   */
```
```suggestion
   // compute the area for each ring in the local basis
```

nyalldawson commented on src/core/geometry/qgscurvepolygon.cpp:
```diff
@@ -492,6 +492,70 @@ double QgsCurvePolygon::area() const
+  const int numRings = static_cast<int>( mInteriorRings.size() + 1 );
+  for ( int ringIndex = 0; ringIndex < numRings; ++ringIndex )
+  {
+    const QgsLineString *ring = static_cast< const QgsLineString * >( ringIndex == 0 ? mExteriorRing.get() : mInteriorRings.at( ringIndex - 1 ) );
```
This should use qgsgeometry_cast, but more importantly -- it will crash for curved geometry types!

nyalldawson commented on src/core/geometry/qgscurvepolygon.cpp:
```diff
@@ -1478,3 +1542,38 @@ int QgsCurvePolygon::compareToSameClass( const QgsAbstractGeometry *other ) cons
+  }
+
+  /*
+   * look for 3 non collinear points on the exterior ring
```
I think this logic is broken -- what would happen if the first 3 points are in the same plane, but then the next point isn't? With the current logic you'll happily return true.

What about rings with a different plane?

Sure, these aren't valid 2.5d geometries, but the function naming and docs misleading suggest that this can be used to test that the polygon IS in a single plane, and it doesn't do that right now...

ptitjano replied:
This is the intended behavior. I have updated the documentation of `has3DPlane` to explain the behavior.

rouault commented:
> compute the cross product of (x, y, z) with a unit vector

EDIT: I mean the dot product, not the cross product

ptitjano committed:
qgsgeometryutils: Introduce CheckWeaklyfor3DPlane

Checks if a 3D geometry has a plane defined by at least 3
non-collinear points.

ptitjano committed:
qgscurve: Implement sumUpArea3D

This is the equivalent of `QgsCurve::sumUpArea()` in 3d dimensions.
The `QgsLineString` uses the 3D shoelace formula whereas the
`QgsCircularString` implementation is based on the projection of the
curve points on a 2D plane and call the 2D version on the projected
points.

ptitjano committed:
testqgspolyhedralsurface: Use valid geometries in tests

ptitjano commented on src/core/geometry/qgscurvepolygon.h:
```diff
@@ -470,6 +472,13 @@ class CORE_EXPORT QgsCurvePolygon: public QgsSurface
 
+    /**
+     * Test if a 3D plane can be extracted from the CurvePolygon.
+     *
```
done

ptitjano commented on src/core/geometry/qgscurvepolygon.cpp:
```diff
@@ -1478,3 +1542,38 @@ int QgsCurvePolygon::compareToSameClass( const QgsAbstractGeometry *other ) cons
+
+  /*
+   * look for 3 non collinear points on the exterior ring
+   */
```
not applicable anymore.

ptitjano commented on src/core/geometry/qgsgeometry.h:
```diff
@@ -488,9 +488,22 @@ class CORE_EXPORT QgsGeometry
+     * \warning QgsGeometry objects are inherently Cartesian/planar geometries, and the area
+     * returned by this method is calculated using strictly Cartesian mathematics.
+     *
+     * \see area()
```
it now raises an exception.

rouault reviewed (commented):
(deleted comment that was meant for another PR...)

rouault commented on tests/src/core/geometry/testqgscircularstring.cpp:
```diff
@@ -1623,6 +1625,48 @@ void TestQgsCircularString::sumUpArea()
+{
+  QgsCircularString cs;
+  double area3D = 1.0; // sumUpArea3D adds to area, so start with non-zero value
+
```
the below results are not obvious to me (not saying they are wrong, but cannot be easily verified immediately).
I would suggest doing an half circle like (0,0),(1 1),(2 0), and rotate it so it fits on the plane of x = 0 and y = 0  (i.e.  (0 0 0),(0 1 1),(0 2 0)   and (0 0 0),(1 0 1),(2 0 2)   and their symetric), so they have all the same areas (with +/- signs)

rouault commented on tests/src/core/geometry/testqgscurvepolygon.cpp:
```diff
@@ -1705,6 +1709,48 @@ void TestQgsCurvePolygon::testCast()
+{
+  QgsCurvePolygon poly;
+
+  std::unique_ptr<QgsCircularString> ext = std::make_unique<QgsCircularString>();
```
same suggestion as above. Test cases with a unit square on planes x = 0 and y = 0 would be great

rouault replied:
and also testing a "U" shape polygon with a mix of convexity and concavity, like (0 0,0 2,1 2,1 1,2 1,2 2,2 0,0 0))

nyalldawson commented on python/PyQt6/core/auto_generated/geometry/qgsgeometry.sip.in:
```diff
@@ -450,6 +450,24 @@ Returns the planar, 2-dimensional area of the geometry.
+
+   QgsGeometry objects are inherently Cartesian/planar geometries, and the area
+   returned by this method is calculated using strictly Cartesian mathematics.
+
```
I'd also add a warning here that this method assumes the horizontal and vertical coordinates are in the same unit.

nyalldawson commented on src/core/geometry/qgslinestring.h:
```diff
@@ -1054,6 +1056,17 @@ class CORE_EXPORT QgsLineString: public QgsCurve
+     * 3D version.
+     *
+     * If the linestring is closed (i.e. a polygon) then the polygon area is equal to the value of the sum.
+     * Please note tha unlike the 2D version, the sign is not taken into account and the returned value is always positive
```
```suggestion
     * Please note that unlike the 2D version, the sign is not taken into account and the returned value is always positive
```

nyalldawson commented on src/core/geometry/qgslinestring.h:
```diff
@@ -1054,6 +1056,17 @@ class CORE_EXPORT QgsLineString: public QgsCurve
+     * Please note tha unlike the 2D version, the sign is not taken into account and the returned value is always positive
+     *
+     * \note If the geometry is 2D, the method falls back to the 2D computation.
+     */
```
Missing \since

ptitjano commented on src/core/geometry/qgscircularstring.cpp:
```diff
@@ -1434,6 +1435,88 @@ void QgsCircularString::sumUpArea( double &sum ) const
+    return;
+  }
+
+  // FIXME: Implement proper 3D shoelace formula for arbitrary 3D polygons
```
indeed. Done

ptitjano commented:
Fix latest remarks + removed some `QStringLiteral` which were introduced in the unit tests.
